[DDP/Functional Optim] Support kwarg arguments

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62079
* #62078
* #61756

Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

rohan-varma committed:
[DDP/Functional Optim] Support kwarg arguments

Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -69,22 +69,20 @@ def decompress(fut):
+        self.functional_optimizer = functional_optim_cls(
             [],
-            *self.functional_optim_args,
+            *functional_optim_args,
```
Optimizers usually only have kwargs. Do you have any example that needs args?

rohan-varma replied:
AFAIK user is able to construct regular optimizer without specifying kwargs and just using the parameter order in constructor:
```
s = torch.optim.SGD(torch.nn.Linear(1,1).parameters(), 0.01)
```

so wanted to support the same behavior here.

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP/Functional Optim] Support kwarg arguments"


Adds support for kwarg arguments into functional optimizer running as
hook.

Differential Revision: [D29838127](https://our.internmc.facebook.com/intern/diff/D29838127/)

[ghstack-poisoned]
